File: leetcode/268-missing-number.py
# https://leetcode.com/problems/missing-number/ (top-interview-questions)

from typing import List

class Solution:
    def missingNumber(self, nums: List[int]) -> int:
        missingNumber = len(nums)
        for i in range(len(nums)):
            missingNumber = missingNumber ^ nums[i] ^ i
        return missingNumber


    def missing_number(self, nums: List[int]) -> int:
        missingNumber = 0
        for i in range(len(nums)+1): missingNumber += i
        missingNumber -= sum(nums)
        return missingNumber

File: leetcode/268-missing-number.js
// https://leetcode.com/problems/missing-number/ (top-interview-questions)

/**
 * @param {number[]} nums
 * @return {number}
 */
var missingNumber = function(nums) {

  let missingNum = nums.length;

  for (let i = 0; i < nums.length; i += 1) {
    missingNum ^= i ^ nums[i];
  }

  return missingNum;

  //-----------------------------------------------other-solution
  let practicleSum = 0;
  let thoraticalSum = nums.length;
  for (let i = 0; i < nums.length; i += 1) {
    thoraticalSum += i;
    practicleSum += nums[i];
  }
  return thoraticalSum - practicleSum;
};
